skimage/morphology/max_tree.py

```diff
@@ -0,0 +1,656 @@
+    components resulting from sequential thresholding operations applied to an
+    image. A connected component at one level is parent of a component at a
+    higher level if the latter is included in the first. A max-tree is an
+    efficient representation of a component tree. A connected component at
+    one level is represented by one reference pixel at this level, which is
+    parent to all other pixels at that level and the reference pixel at the
+    level above. The max-tree is the basis for many morphological operators,
+    namely connected operators.
+
+    Parameters
+    ----------
+    img: ndarray
+        The input image for which the max-tree is to be calculated.
+        This image can be of any type.
+    connectivity: unsigned int, optional
+        The neighborhood connectivity. The integer represents the maximum
+        number of orthogonal steps to reach a neighbor. It is 1 for
+        4-connectivity and 2 for 8-connectivity. Default value is 2.
+
+    Returns
+    -------
+    parent: ndarray, int64
+        The value of each pixel is the index of its parent in the ravelled
+        array
+
+    tree_traverser: 1D array, int64
+        The ordered pixel indices (referring to the ravelled array). The pixels
+        are ordered such that every pixel is preceded by its parent (except for
+        the root which has no parent).
+
+    References
+    ----------
+    .. [1] Salembier, P., Oliveras, A., & Garrido, L. (1998). Antiextensive
+           Connected Operators for Image and Sequence Processing.
+           IEEE Transactions on Image Processing, 7(4), 555-570.
+    .. [2] Berger, C., Geraud, T., Levillain, R., Widynski, N., Baillard, A.,
+           Bertin, E. (2007). Effective Component Tree Computation with
+           Application to Pattern Recognition in Astronomical Imaging.
+           In International Conference on Image Processing (ICIP) (pp. 41-44).
+    .. [3] Najman, L., & Couprie, M. (2006). Building the component tree in
+           quasi-linear time. IEEE Transactions on Image Processing, 15(11),
+           3531-3539.
+    .. [4] Carlinet, E., & Geraud, T. (2014). A Comparative Review of
+           Component Tree Computation Algorithms. IEEE Transactions on Image
+           Processing, 23(9), 3885-3895.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from skimage.max_tree import build_max_tree
+
+    We create a small sample image (Figure 1 from [4]) and build the max-tree.
+
+    >>> image = np.array([[15, 13, 16], [12, 12, 10], [16, 12, 14]])
+    >>> P, S = build_max_tree(image, connectivity=2)
+
+    """
+    # User defined masks are not allowed, as there might be more than one
+    # connected component in the mask (and therefore not a single tree that
+    # represents the image). Mask here is an image that is 0 on the border
+    # and 1 everywhere else.
+    mask_shrink = np.ones([x - 2 for x in image.shape], bool)
+    mask = np.pad(mask_shrink, 1, mode='constant')
```

Comment:
I believe that this will be much slower than creating `np.ones` of the correct size and then setting all the borders to 0. You can do this by calling `np.rollaxis` repeatedly and setting `arr[0] = 0` and `arr[-1] = 0`.

Reply:
That's nice 👍